optimizer/plan/cost.go

```diff
@@ -117,33 +87,9 @@ func (c *costEstimator) tableScan(v *TableScan) {
 }
 
 func (c *costEstimator) indexScan(v *IndexScan) {
-	var rowCount float64
-	if len(v.Ranges) == 1 && v.Ranges[0].LowVal[0] == nil && v.Ranges[0].HighVal[0] == MaxVal {
-		// full range use default row count.
-		rowCount = FullRangeCount
-	} else {
-		for _, v := range v.Ranges {
-			// for condition like 'a = 0'.
-			if v.IsPoint() {
-				rowCount++
-				continue
-			}
-			// For condition like 'a < 0'.
-			if v.LowVal[0] == nil || v.LowVal[0] == MinNotNullVal {
-				rowCount += HalfRangeCount
-			}
-			// For condition like 'a > 0'.
-			if v.HighVal[0] == MaxVal {
-				rowCount += HalfRangeCount
-			}
-			// For condition like 'a > 0 and a < 1'.
-			rowCount += MiddleRangeCount
-		}
-		// If the index has too many ranges, the row count may exceed the default row count.
-		// Make sure the cost is lower than full range.
-		if rowCount >= FullRangeCount {
-			rowCount = FullRangeCount - 1
-		}
+	var rowCount float64 = FullRangeCount
+	for _, con := range v.AccessConditions {
+		rowCount *= guesstimateFilterRate(con)
```

Comment:
MySQL uses `guesstimate` for this kind of rule.
Estimate is based on index statistics.